### PR TITLE
CA1002: Do not expose generic lists

### LINKED ISSUE #369: Port FxCop rule CA1002: DoNotExposeGenericLists
**Title:** Do not expose generic lists

**Description:**

System.Collections.Generic.List&lt;(Of &lt;(T&gt;)&gt;) is a generic collection that is designed for performance, not inheritance. Therefore, List does not contain any virtual members. The generic collections that are designed for inheritance should be exposed instead.

**Dependency:** None

**Notes:**

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #369

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericListsTests.cs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        public async Task CA1002_ExtensionMethods()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System.Collections.Generic;
+
+namespace Namespace1
+{
+    public static class Helper
+    {
+        public static bool Ext1(this List<int> list) => true;
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericListsTests.cs]: This exclusion is not part of the original implementation but that seems weird to get a penalty to extend `List<T>` so I added an exclusion. Please feel free to let me know if you think this shouldn't be done.

**Comment by mavasani** [comment]
FYI: You will need to fix a new RS diagnostic once https://github.com/dotnet/roslyn-analyzers/pull/3444 goes in.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(context =>
+            {
+                var genericListType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericList1);
+
+                context.RegisterSymbolAction(context =>
+                {
+                    var symbol = context.Symbol;
+                    var methodSymbol = symbol as IMethodSymbol;
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:45]: Did you mean to register separate actions for each symbol kind? It seems better to move the shared code in a separate helper and register separate actions for each symbol kind that needs different checks.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(context =>
+            {
+                var genericListType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericList1);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:40]:
```suggestion
                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericList1, out var genericListType))
                {
                   return;
                }
```

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+                context.RegisterSymbolAction(context =>
+                {
+                    var symbol = context.Symbol;
+                    var methodSymbol = symbol as IMethodSymbol;
+
+                    // FxCop compat: only analyze externally visible symbols by default.
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:47]: Thanks!

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:47]: Please update https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#analyzed-api-surface and file a doc issue to mention api surface configurability for this rule. Same for other rules that you are supporting this option.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs (hunk 2 above)
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:45]: Do you mean replacing 

```csharp
context.RegisterSymbolAction(ctx => {}, SymbolKind.Field, SymbolKind.Property, SymbolKind.Method);
``` 

with 

```csharp
context.RegisterSymbolAction(Analyze, SymbolKind.Field);
context.RegisterSymbolAction(Analyze, SymbolKind.Property);
context.RegisterSymbolAction(Analyze, SymbolKind.Method);

private static Analyze() {}
```

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:45]: I meant having separate `AnalyzeMethod`, `AnalyzeField`, etc. for each symbol kind, so you can directly cast to `IMethodSymbol` in `AnalyzeMethod` and have method specific logic only in `AnalyzeMethod`.

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotExposeGenericLists.cs:45]: Ah ok, yeah makes sense I have overused the helpers :)

**Pull request merged**